### DISCUSSION TITLE
How to test if the request does contain the expected header?

### DISCUSSION THREAD

**Discussion by marspox** [reactiveui/refit#1521] [Q&A] [answered]
Hello,
Is there any way, to get HttpRequestMessage and check if it contains a specific header for instance UserAgent?
I would like to cover my code by the test to be sure that the sent UserAgent value is the expected one, currently, I did it by DelegatingHandler, but I'm wondering if there exists a better option since the current one looks ugly to me.

Thank you in advance, for any insights.

**Comment by glennawatson** [comment, marked as answer]
Yes. Use a `DelegatingHandler` to assert on the outgoing `HttpRequestMessage`, that is the supported and idiomatic approach for verifying headers, URL, body, etc. It is not "ugly" - it is exactly the seam .NET gives you for inspecting requests.

A clean pattern is a small test handler that captures the request:

```csharp
sealed class CapturingHandler : DelegatingHandler
{
    public HttpRequestMessage? Request { get; private set; }

    protected override Task<HttpResponseMessage> SendAsync(
        HttpRequestMessage request, CancellationToken ct)
    {
        Request = request;
        return Task.FromResult(new HttpResponseMessage(HttpStatusCode.OK));
    }
}
```

Wire it up and assert:

```csharp
var handler = new CapturingHandler();
var client = new HttpClient(handler) { BaseAddress = new Uri("https://example.com") };
var api = RestService.For<IMyApi>(client);

await api.GetThing();

Assert.Equal("my-agent", handler.Request!.Headers.UserAgent.ToString());
```

If you want less boilerplate, libraries like RichardSzalay.MockHttp give you a fluent `.WithHeaders(...)` matcher over the same `DelegatingHandler` mechanism. Either way, inspecting `HttpRequestMessage` via a handler is the correct option - there is no separate Refit API for this.
